extensions/windows-event-log/Bookmark.cpp

```diff
@@ -0,0 +1,217 @@
+      return;
+    }
+  }
+
+  ok_ = true;
+}
+
+Bookmark::~Bookmark() {
+  if (file_.is_open()) {
+    file_.close();
+  }
+
+  if (hBookmark_) {
+    EvtClose(hBookmark_);
+  }
+}
+
+Bookmark::operator bool() {
+  return ok_;
+}
+  
+bool Bookmark::hasBookmarkXml() {
+  return !bookmarkXml_.empty();
+}
+
+EVT_HANDLE Bookmark::bookmarkHandle() {
+  return hBookmark_;
+}
+
+bool Bookmark::saveBookmark(EVT_HANDLE hEvent)
+{
+  if (!EvtUpdateBookmark(hBookmark_, hEvent)) {
+    logger_->log_error("!EvtUpdateBookmark error: %d.", GetLastError());
+    return false;
+  }
+
+  // Render the bookmark as an XML string that can be persisted.
+  DWORD bufferSize{};
+  DWORD bufferUsed{};
+  DWORD propertyCount{};
+  if (!EvtRender(0, hBookmark_, EvtRenderBookmark, bufferSize, 0, &bufferUsed, &propertyCount)) {
+    DWORD status = ERROR_SUCCESS;
+    if (ERROR_INSUFFICIENT_BUFFER == (status = GetLastError())) {
+      bufferSize = bufferUsed;
+
+      std::vector<wchar_t> buf(bufferSize / 2 + 1);
+
+      if (!EvtRender(0, hBookmark_, EvtRenderBookmark, bufferSize, &buf[0], &bufferUsed, &propertyCount)) {
+        logger_->log_error("!EvtRender error: %d.", GetLastError());
+        return false;
+      }
+
+      file_.seekp(std::ios::beg);
+
+      // Write new bookmark over old and in the end write '!'. Then new bookmark is read until '!'.
+      file_ << &buf[0] << L'!';
+
+      file_.flush();
+    }
+    else if (ERROR_SUCCESS != (status = GetLastError())) {
+      logger_->log_error("!EvtRender error: %d.", GetLastError());
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool Bookmark::createEmptyBookmarkXmlFile() {
+  if (file_.is_open()) {
+    file_.close();
+  }
+
+  file_.open(filePath_, std::ios::out);
+
+  auto ret = file_.is_open();
+  if (!ret) {
+    logger_->log_error("Cannot open %s", filePath_.c_str());
+  }
+
+  return ret;
+}
+
+// Creates directory "processor_repository\ConsumeWindowsEventLog\uuid\{uuid}" under "bin" directory.
+bool Bookmark::createUUIDDir(const std::string& uuid, std::string& dir)
+{
+  auto getRootDirectory = [](std::string& rootDir, std::shared_ptr<logging::Logger> logger)
+  {
+    char dir[MAX_PATH + 1];
+    auto length = GetModuleFileName(0, dir, _countof(dir) - 1);
+    if (!length) {
+      logger->log_error("!GetModuleFileName error: %x", GetLastError());
+      return false;
+    }
+    dir[length] = '\0';
+
+    auto pBackslash = strrchr(dir, '\\');
+    if (!pBackslash) {
+      logger->log_error("!pBackslash");
+      return false;
+    }
+    *(pBackslash + 1) = '\0';
+
+    rootDir = dir;
+
+    return true;
+  };
+
+  dir.clear();
+
+  if (!getRootDirectory(dir, logger_))
+    return false;
+
+  for (const auto& curDir : std::vector<std::string>{ "processor_repository", "ConsumeWindowsEventLog", "uuid", uuid }) {
+    dir += curDir + '\\';
+
+    if (GetFileAttributes(dir.c_str()) == INVALID_FILE_ATTRIBUTES) {
```

Comment:
in FileUtils, there is a function called "create_dir".
It was just fixed for windows and returns true in case the dir could be created or it already exists. 
I think that's exactly what you need here.

Reply:
Fixed.